### PR TITLE
[DL Edition] T038: Protein Ligand Interaction Prediction

### PR DESCRIPTION
**Pull request by Old-Shatterhand**
## Description
Proof of concept for GNN-based protein ligand interaction prediction in talktorial T038

## Todos
- [x] Find an appropriate dataset
- [x] Fix bug in code to be able to train on batches larger than one sample

## Questions
None

## Status
Initial draft for further discussion

### CODE REVIEW

**Comment by Old-Shatterhand** [comment]
## Pull request update

After some minor updates on the nodebook, the following work is left.

## TODOS:
- [ ] Find an appropriate dataset for training
- [ ] Discussion: Put plots and explanations of training based on selected dataset
- [ ] Questions: Come up with questions about the content

**Comment by Old-Shatterhand** [comment]
# Talktorial review

Pullrequest of talktorial about GNN-based protein-ligand interaction prediction.

## Details

* Talktorial ID: T038
* Title: Protein-Ligand Interaction Prediction
* Original authors: Roman Joeres
* Reviewer(s): Andrea Volkamer, tbd
* Date of review: DD-MM-YYYY (tbd)

## Content

* One line summary: Introduction of GNN-based protein-ligand interaction prediction
* Potential labels or categories (e.g. machine learning, small molecules, online APIs): Kinases, Machine Learning, Graph Neural Networks
* Time it took to execute (approx.): 1 hour
* [ ] I have used the [talktorial template](https://github.com/volkamerlab/teachopencadd/blob/master/teachopencadd/talktorials/T000_template/talktorial.ipynb) and followed the content and formatting suggestions there
* [ ] _Packages_ must be open-sourced and should be installable from `conda-forge`. If you are adding new packages to the TeachOpenCADD environment, please check if already installed packages can perform the same functionality and if not leave a sentence explaining why the new addition is needed. If the new package is not on `conda-forge`, please list them and their intended usage here.
  * `biotite`, `pypdb`, `chembl-webresource-client`, `rdkit`: Already in TeachOpenCADD
  * `torch 1.10.1`, `torch-geometric 2.2.0`, `torch-cluster 1.6.0`, `torch-scatter 2.0.9`, `torch-sparse 0.6.13`, `torch-spline-conv 1.2.1`, `cpuonly 2.0`: From "dubious" sources (either own conda-channel (`pyorch`) or installed as pip wheel. All of them only for cpu
* [ ] _Data_ must be publicly available, preferably accessible via a webserver or downloadable via a URL. Please list the data resources that you use and how to access them:
  * KiBA dataset: Access via [website](https://researchportal.helsinki.fi/fi/datasets/kiba-a-benchmark-dataset-for-drug-target-prediction)

## Content style

* [ ] Talktorial includes cross-references to other talktorials if applicable
* [ ] The table of contents reflects the talktorial story-line; order of #, ##, ### headers is correct
* [ ] URLs are linked with meaningful words, instead of pasting the URL directly or linking words like `here`.
* [ ] I have spell-checked the notebook
* [ ] Images have enough resolution to be rendered with quality, without being _too_ heavy.
* [ ] All figures have a description
* [ ] Markdown cell content is still in-line with code cell output (whenever results are discussed)
* [ ] I have checked that cell outputs are not incredibly long (this applies also to `DataFrames`)
* [ ] Formatting looks correctly on the Sphinx render (bold, italics, figure placing)

## Code style

* [ ] Variable and function names follow snake case rules (e.g. `a_variable_name` vs `aVariableName`)
* [ ] Spacing follows PEP8 (run Black on the code cells if needed)
* [ ] Code line are under 99 characters each (run `black-nb -l 99`)
* [ ] Comments are useful and well placed
* [ ] There are no unpythonic idioms like `for i in range(len(list))` (see slides)
* [ ] All 3rd party dependencies are listed at the top of the notebook
* [ ] I have marked all code cell with output referenced in markdown cells with the label `# NBVAL_CHECK_OUTPUT`
* [ ] I have identified potential candidates for a code refactor / useful functions
* [ ] All `import ...` lines are at the top (practice part) cell, ordered by standard library / 3rd party packages / our own (`teachopencadd.*`)
* [ ] I have used absolute paths instead of relative paths
  ```python
  HERE = Path(_dh[-1])
  DATA = HERE / "data"
  ```

## Website
We present our talktorials on our TeachOpenCADD website (https://projects.volkamerlab.org/teachopencadd/), so we have to check as well if the Jupyter notebook renders nicely there.

* [ ] If this PR adds a new talktorial, please follow these steps:
  * [ ] Add your talktorial to the complete list of talktorials [here](https://github.com/volkamerlab/teachopencadd/blob/master/docs/all_talktorials.rst) (at the end).
  * [ ] Add your talktorial to one or multiple of the collections [here](https://github.com/volkamerlab/teachopencadd/blob/master/docs/talktorials.rst). Or propose a new collection section in your PR.
  * [ ] Add your talktorial's `nblink` file by running `python generate_nblinks.py` from within the directory `teachopencadd/docs/talktorials`.
  * [ ] Please complile the website following the instructions [here](https://github.com/volkamerlab/teachopencadd/tree/master/docs).
* [ ] Check the rendering of the talktorial of this PR.
* [ ] Is your talktorial listed [in the talktorial list](https://projects.volkamerlab.org/teachopencadd/all_talktorials.html)?
* [ ] Is your talktorial listed [in the talktorial collections](https://projects.volkamerlab.org/teachopencadd/talktorials.html)?
  * [ ] Add a picture for your talktorial in the collection view by following [these instructions](https://github.com/volkamerlab/teachopencadd/discussions/185).

**Comment by gerritgr** [comment]
- I will (and elsewhere)-> We will? (I think in other notebooks, it is third person, have not checked though).
- "field of protein ligand interaction prediction" -> "protein-ligand"
- Maybe explain the terms protein and ligand very shortly in the intro. 
- Titles are lowercase (Sentence Case) in the other notebooks. 
- "..., I'll link to this otherwise, I'll explain new things below." -> "... , I will link to this. Otherwise, I will explain new things below." 
- "one wants to" -> wants
- There are some other typos, but Grammarly or so can catch these I guess.
- simple Feed-forward Neural Network (FNN): do you mean MLP? GNNs are also technically feed-forward networks.
- State in the beginning of the workflow that this is a binary classification task.
- " from the PDB entry with ID 4O75." link to a talkturial explaining PDB earlier.
- explain C_alpha 
- In the "Technical background", say that we use the same GNN architecture for both ligands and proteins.
- I think the BCE explanation should be clarified. What are the negative and positive samples here (binding and non-binding?)?
- suppress DtypeWarning and add a comment explaining what kiba_preprocessing is doing
- "Storing and representing data in PLI-prediction is a bit different from other neural networks. ": You mean the input data (i.e., the graphs), right? Maybe clarify this. Also, even though it is obvious, PLI was not introduced as an abbreviation. 
- Maybe merge together with the "Data Points" subsection. 
- I have not checked the code yet.

**Comment by Old-Shatterhand** [comment]
I implemented Gerrits comments and uploaded a new notebook.

File: teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb [hunk 1]
```diff
@@ -0,0 +1,995 @@
+{
```
Comment (mbackenkoehler) [inline teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb:1]: <ul><li>Describe interaction a bit more. Is it only about binding? If so maybe mention classical approaches to the same problem.</li><li>For example (missing comma)</li><li>You could motivate the model with virtual screening of compound libraries, for example</li><li>TBC in the end.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/290/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (mbackenkoehler) [inline teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb:1]: <ul><li>Maybe change workflow to 'model' something similar; The workflow would include data prep, training and so on, I presume.</li><li>Simplify the second sentence.</li><li>"We will only use the information if an interaction exists or not," -&gt; This sounds a bit strange. Just say right away, you are transforming the task to a classification</li><li>introduction of FNN acronym is missing</li><li>link to rcsb.org for 4O75</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/290/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (mbackenkoehler) [inline teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb:1]: <ul><li>capitalization in Binary Cross Entropy?</li><li>The loss is</li><li>introduce BCE acronym</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/290/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (mbackenkoehler) [inline teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb:1]: <ul><li>use Path library in last line; maybe put filepaths in </li><li>KeyboardInterrupt in output</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/290/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (mbackenkoehler) [inline teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb:1]: <ul><li>as it's -&gt; its, from -&gt; in</li><li>could have used an enum? ;)</li><li>raise on invalid molecules?</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/290/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (mbackenkoehler) [inline teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb:1]: <ul><li>maybe link a PDB spec as an explanation</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/290/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (mbackenkoehler) [inline teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb:1]: <ul><li>Data class -&gt; data class</li><li>link pytorch geometric?</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/290/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (mbackenkoehler) [inline teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb:1]: <ul><li>read in -&gt; read</li><li>process it / process them (and data -&gt; data points)</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/290/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (mbackenkoehler) [inline teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #25.            with open(os.path.join(self.folder_name, "tables", "ligands.tsv"), "r") as data:
</pre><p>use path library for consistency</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/290/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (mbackenkoehler) [inline teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #32.                (filename[:-4], pdb_to_graph(os.path.join(os.path.join(self.folder_name, "proteins", filename)))) for filename in os.listdir(os.path.join(self.folder_name, "proteins"))
</pre><p>pathlib (see above)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/290/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (mbackenkoehler) [inline teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #62.                # then split the data and store them for later reuse without running the preprocessing pipeline
</pre><ul><li>consider a torch data splitter?</li></ul><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/290/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (mbackenkoehler) [inline teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    class Encoder(torch.nn.Module):
</pre><p>Maybe use encoding in fig 1 for consistency</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/volkamerlab/teachopencadd/pull/290/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

[the remark this answers is not in the document]
Comment (Old-Shatterhand) [reply teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb:1]: <p>Regarding second point: I guess, one-hot is the solution, Grammarly would suggest. I changed the text occurrences to "one-hot".</p>

Comment (Old-Shatterhand) [reply teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb:1]: <p>First point: In the PyTorch docu, BCE is written in camel case.</p><p>Second point: Where?</p>

Comment (Old-Shatterhand) [reply teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb:1]: <p>Yeah, enum would be an option, but another programming concept where is simply need a mapping. One can do it.</p><p>Regarding the "raise on invalid molecules": In principle yes, but I don't want to raise an error here and catch it below and make the whole code more complex.</p>

Comment (Old-Shatterhand) [reply teachopencadd/talktorials/T038_protein_ligand_interaction_prediction/talktorial.ipynb:1]: <p>Would be too complex here, I'd say. Here, you easily see what's happening and how and why.</p>
